Fixes #27724 - Add a unique-id setting to foreman.

ShimShtein opened the pull request:
This introduces a persisted randomly generated identifier of the Foreman
instance. This is useful for correlating foreman-debug tarballs with 
logs and
other debugging data that users upload to developers when they need 
help. This
can be also useful to identify multiple instances in external systems, 
e.g.
when workflows across multiple Foremans are orchestrated.

ares commented on app/models/setting/general.rb:
```diff
@@ -26,6 +26,7 @@ def self.default_settings
       self.set('lab_features', N_("Whether or not to show a menu to access experimental lab features (requires reload of page)"), false, N_('Show Experimental Labs')),
       self.set("append_domain_name_for_hosts", N_("Foreman will append domain names when new hosts are provisioned"), true, N_("Append domain names to the host")),
       self.set('outofsync_interval', N_("Duration in minutes after servers are classed as out of sync."), 30, N_('Out of sync interval')),
+      self.set('instance_id', N_("Foreman instance ID, uniquely identifies this Foreman instance."), Foreman.uuid, N_('Foreman UUID')),
```
This changes the value with every restart, since settings is set to nil and we use the default value. We may want to generate it in migration.

ShimShtein replied:
It doesn't - on first run it creates a record in `settings` table, and the value will stay there. At least that's what I see in my dev DB.

ares replied:
that didn't happen on my DB when I tested, I think we don't save any value to DB until the setting is changed, nil means use the default value which in this case change with every restart

ShimShtein replied:
@ares, my bad, didn;t count the parameters right. Now I set `value` instead of `default_value` which is not overridden after multiple `rails s` invocations.

ares reviewed (changes requested):
Thansk @ShimShtein, left one comment + it would also be nice to add to rss user agent at https://github.com/theforeman/foreman/blob/f2ee562ece5d5084f39160524ac57b23b7bde251/app/services/ui_notifications/rss_notifications_checker.rb#L97

ShimShtein commented:
> it would also be nice to add to rss user agent

Talking to @GregSutcliffe confirmed my thoughts about this being a bit too intrusive.
From looking at google search results, you see people trying to get rid of user-agent fingerprinting as much as possible. Adding a clear way for identification seems a bit of bad publicity, even if it's a well known endpoint (rss feed).

From our conversation:
> it's much easier to argue for aggregate-only stats when you're not collecting UUIDs

ares commented:
Ok, I don't insist, it just feels we get wrong metrics from LB/HA setups. Also it would be easy to blacklist specific UUIDs that e.g. runs in CI. @GregSutcliffe any further thoughts on this? Could you be more specific, what the UUID adds comparing to IP address that we already know? I mean the correlation is already possible by IP.

But as I said, I don't insist at all, just felt as a good addition.

ekohl commented:
I'd like to see some argumentation of why this is needed.

ShimShtein commented:
@ekohl The TLDR version is to use it as subscriber ID for any service where the ID should survive restart. For example any publication originated on SaaS to a specific instance.

Edit: tagged @ekohl.

lzap commented:
I think this does not belong to general settings but the YAML file, this should be deployed by the installer and properly documented with an easy opt-out option which would set this to `nil` or blank.

UUID itself does not make any sense, we need to make some use of it - RSS is a good thing for start

ekohl commented:
Putting this in the YAML file makes deployment harder in clusters. Doing this as part of the seeding is probably the correct place.

I do share the concerns that this alone doesn't appear very useful, but RSS also introduces some privacy concerns.

> The TLDR version is to use it as subscriber ID for any service where the ID should survive restart. For example any publication originated on SaaS to a specific instance.

This sounds like an odd case. I'd like to hear the non-TLDR version :) In particular, usually there's some authentication to an external service. Exceptions to this is tracking and typically sysadmins dislike this.

@lzap would this make sense in the telemetry part that's in core? That is opt-in and could make sense to users without any privacy concerns.

timogoebel commented:
> I'd like to hear the non-TLDR version :) In particular, usually there's some authentication to an external service.

Yep, I don't get the use case as well. What are we trying to achieve with this?

lzap commented:
> would this make sense in the telemetry part that's in core?

Nope, telemetry is local deployment.

ares commented:
The use case for this is: there's a new plugin that sends out inventory data somewhere else, cloud.redhat.com in this case. We plan to add a receptor service (optional) next to Foreman, that allows those "somewhere elses" connect back to Foreman through the receptor and to interact with hosts managed through Foreman. Given the external inventory can contain information about multiple Foremans, we need to know, which Foreman we should connect to. Hence, we need unique identifier.

The reason to put it into the core was, it's useful for many other cases
1) makes it easier to distinguish statistics we got from RSS
2) we planned to integrate pendo service before, which also needed the same unique differentiator
3) the use case above is now limited to one feature (executing insights plan remediation) but I'm sure we'll get more in future

The reason why settings and not YAML file or other thing
1) good for LB/HA setups, where it's in single DB and multiple instances represent one Foreman
2) works well with backup/restore
3) people can see it (needed when you want to use it somewhere) and fetch it through API
4) people can change it (HA setup, debugging) and therefore also make it not unique and "opt-out"

I think it does not add any additional tracking trail even if we add it to RSS user agent. The UUID is random and is not connected to any other DB information. But I may be missing something, so I'm fine not adding it to RSS if that raises someones eyebrow. We need this in 1.24, we can start the discussion on discourse with users, but I'd be happier if we don't block the merge on it, as there's opt-out option.

lzap commented:
> pendo service

What's that?

> why settings and not YAML

These are good reasons, I agree that Settings is a good place. Setting it to blank value should be advertised as opt-out.

ekohl commented:
> The use case for this is: there's a new plugin that sends out inventory data somewhere else, cloud.redhat.com in this case. We plan to add a receptor service (optional) next to Foreman, that allows those "somewhere elses" connect back to Foreman through the receptor and to interact with hosts managed through Foreman. Given the external inventory can contain information about multiple Foremans, we need to know, which Foreman we should connect to. Hence, we need unique identifier.

I don't see how a UUID is useful here. There should still be authentication to know which cloud.redhat.com account it is. If you want to identify multiple installations, shouldn't it be stored next to that?

Have you considered people who clone their installation to a lab? Would that be an issue?

Would it make sense to derive unique installations from the `foreman_url` setting?

ares commented:
> > pendo service
> 
> What's that?

Google analytics like thing. It was supposed to tell us, what pages people use most. It got deprioritized and was supposed to be opt-in. Anyway, unrelated to this discussion.

ares commented:
> I don't see how a UUID is useful here. There should still be authentication to know which cloud.redhat.com account it is. If you want to identify multiple installations, shouldn't it be stored next to that?

Imagine there are open channel between foreman instances and cloud.redhat.com. cloud.redhat.com decides to contact host1,host2 for which it knows it's handled by foreman1 (unique identifier), the conncetion is using that channel that we need to identify somehow. Foreman does not have to be reachable directly, that's why we have these channels provided by receptor. We just need to know which channel to use. It's one channel per foreman. Channel id = foreman id.

> Have you considered people who clone their installation to a lab? Would that be an issue?

not sure what you mean by that, multiple foreman behind one IP? if that's the case, it's fine, each lab instance has unique channel. The channel is opened from foreman side, but that can be used for push communications from the other side.

> Would it make sense to derive unique installations from the foreman_url setting?

I don't think so, I have multiple foreman.example.com instances

ekohl commented:
> Imagine there are open channel between foreman instances and cloud.redhat.com. cloud.redhat.com decides to contact host1,host2 for which it knows it's handled by foreman1 (unique identifier), the conncetion is using that channel that we need to identify somehow. Foreman does not have to be reachable directly, that's why we have these channels provided by receptor. We just need to know which channel to use. It's one channel per foreman. Channel id = foreman id.

That sounds like the receptor needs to tag channels. Another way of solving that is generating credentials per satellite instance on the cloud.redhat.com side. That also allows revoking access from the cloud.redhat.com side. Never trust user generated data is a common rule since it could easily be used to generate an excessive amount of installations.

ares commented:
> That sounds like the receptor needs to tag channels. Another way of solving that is generating credentials per satellite instance on the cloud.redhat.com side. That also allows revoking access from the cloud.redhat.com side. Never trust user generated data is a common rule since it could easily be used to generate an excessive amount of installations.

I'm not sure how credentials are related to this. Credentials all by itself wouldn't be matched to host's Foreman instance that is uploaded as part of inventory upload. We'd still need some ID that we can associate hosts with. That would be in this case credentials ID? Hence just a different ID, this time not even random baesd.

Maybe it will be easier to draw a picture when I have a chance to talk to you face to face. But I still don't understand, what is the concern with adding randomly generated instance ID.

ezr-ondrej commented on app/models/setting/general.rb:
```diff
@@ -26,6 +26,7 @@ def self.default_settings
       self.set('lab_features', N_("Whether or not to show a menu to access experimental lab features (requires reload of page)"), false, N_('Show Experimental Labs')),
       self.set("append_domain_name_for_hosts", N_("Foreman will append domain names when new hosts are provisioned"), true, N_("Append domain names to the host")),
       self.set('outofsync_interval', N_("Duration in minutes after servers are classed as out of sync."), 30, N_('Out of sync interval')),
+      self.set('instance_id', N_("Foreman instance ID, uniquely identifies this Foreman instance."), 'uuid', N_('Foreman UUID'), Foreman.uuid),
```
Is it actually desirable for users to change it?

ares replied:
Yes, in case of LB setup, where 2 instances should act as one. Also that allows to easily opt-out in case people don't want it to be set.

timogoebel replied:
I'd argue, that in a LB setup the instances share the same database. E.g. the instance id is automatically the same for all lb targets. I also think that we should not allow an opt-out of the instance ID. As a developer, I want to rely on having the instance ID being always available. If we send the ID is part of some tracking callback and should provide a way to opt out.
If we decide to make the instance ID available in core, people should only be allowed to change it via a rake task that asks for confirmation imho. This should not be an arbitrary text field.

ares replied:
You're right about the DB so instances will share the UUID anyway. Given the feedback we got here, we slowed down in using this for any kind of tracking (rss user-agent header) and rather will be more careful, allowing people to opt-out by setting any value. If we'll go ahead and will start using this in other usecases too, I think it would be good to introduce read-only settings. This would be handy for other settings. I suppose the question arised - why settings if they can't change it.

The answer would be, we already have this. Settings stored in yaml files are displayed in UI but can be modified. That is useful to easily figure out the value, perhaps even get it via API. The benefit in this case would be, it's stored in DB instead of a file, which is easier to handle in LB setup. If we decide in future to lock it, then I'd agree it would be also handy to have resetting or changing rake task.

Is the current approach acceptable now? I spoke with @ekohl and we agreed to improve commit message which addresses his concerns. Hopefully this would lead to consensus.

timogoebel replied:
Makes sense. If we don't use it for tracking, I'm good. Let's provide a (tiny) proper API to consume this. So we can change the implementation in the future.

ares replied:
Just to be clear, you mean some developer API like ˋForeman.instance_id(=)`? I like the idea. I may even add Jail object/template macro for it (in separate PR ). @ShimShtein, could you please add it to the PR? I think we'd be ready after it.

ShimShtein committed:
Fixes #27724 - Add a unique-id setting to foreman.

This introduces a persisted randomly generated identifier of the Foreman
instance. This is useful for correlating foreman-debug tarballs with 
logs and
other debugging data that users upload to developers when they need 
help. This
can be also useful to identify multiple instances in external systems, 
e.g.
when workflows across multiple Foremans are orchestrated.

ShimShtein commented:
@ares, added `Foreman.unique_id` getter and setter.

timogoebel commented on lib/foreman.rb:
```diff
@@ -22,4 +22,12 @@ def self.in_rake?(*rake_tasks)
     in_rake?('db:create', 'db:migrate', 'db:drop')
   end
+
+  def self.instance_id
```
Good stuff, that‘s what I had in mind.

ares reviewed (approved):
Thanks @ShimShtein for quickly adding accessors, I tested and it works great. I believe all comments here were addressed, the test failure on mysql is unrelated. Merging!
